crates/dekoder/src/inventory.rs

```diff
@@ -0,0 +1,84 @@
+//! Assets manager.
+use glob::glob;
+use lz4_flex::frame::FrameDecoder;
+use ndarray::Array4;
+use ndarray_npy::NpzReader;
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::fs::{read_to_string, File};
+use std::io::{BufReader, Cursor};
+use std::path::PathBuf;
+use yaml_rust2::{Yaml, YamlLoader};
+
+use crate::{EKOError, Operator, Result};
+
+/// Headers are in yaml files.
+const HEADER_EXT: &str = "*.yaml";
+
+/// Assets manager.
+pub(crate) struct Inventory<K: Eq + for<'a> TryFrom<&'a Yaml, Error = EKOError>> {
+    /// Working directory
+    pub(crate) path: PathBuf,
+    /// Available keys
+    pub(crate) keys: HashMap<OsString, K>,
+}
+
+impl<K: Eq + for<'a> TryFrom<&'a Yaml, Error = EKOError>> Inventory<K> {
+    /// Load all available entries.
+    pub fn load_keys(&mut self) -> Result<()> {
+        let path = self.path.join(HEADER_EXT);
+        let path = path
+            .to_str()
+            .ok_or(EKOError::KeyError("due to invalid path".to_owned()))?;
+        for entry in glob(path)
+            .map_err(|_| EKOError::KeyError("because failed to read glob pattern".to_owned()))?
+            .filter_map(core::result::Result::ok)
+        {
+            let cnt = YamlLoader::load_from_str(&read_to_string(&entry)?)
+                .map_err(|_| EKOError::KeyError("because failed to read yaml file.".to_owned()))?;
+            self.keys.insert(
+                entry
+                    .file_name()
+                    .ok_or(EKOError::KeyError(
+                        "because failed to read file name".to_owned(),
+                    ))?
+                    .to_os_string(),
+                K::try_from(&cnt[0])?,
+            );
+        }
+        Ok(())
+    }
+
+    /// List available keys.
+    pub fn keys(&self) -> Vec<&K> {
+        let mut ks = Vec::new();
+        for k in self.keys.values() {
+            ks.push(k);
+        }
+        ks
+    }
+
+    /// Check if `k` is available (with given precision).
+    pub fn has(&self, k: &K) -> bool {
+        self.keys.iter().any(|it| (it.1).eq(k))
+    }
+
+    /// Load `k` from disk.
+    pub fn load(&self, k: &K, v: &mut Operator) -> Result<()> {
+        // Find key
+        let k = self.keys.iter().find(|it| (it.1).eq(k));
+        let k = k.ok_or(EKOError::KeyError("because it was not found".to_owned()))?;
+        let p = self.path.join(k.0).with_extension("npz.lz4");
+        // Read npz.lz4
+        let mut reader = BufReader::new(FrameDecoder::new(BufReader::new(File::open(&p)?)));
+        let mut buffer = Vec::new();
+        std::io::copy(&mut reader, &mut buffer)?;
```

Comment:
As for the one above, most likely you don't even need this: just pass the variable returned by `FrameDecoder::new()` to `NpzReader::new()`.

The idea is that readers are not reading immediately, they are a promise that they will read when needed, i.e. they are lazy - only when you will trigger the read with an operation they will actually do the job.
Here you're triggering it twice: with `copy()`, and with `.by_name()`. But you only need the second, and that's why you're essentially undoing `copy()` by wrapping it in a `Cursor` (that will make it similar again to a reader, i.e. a `Read` implementor, by wrapping a memory buffer).

Reply:
... while I could remove `BufRead`, I can not drop the copy to `Vec` - there is a type mismatch (`FrameDecoder` has no `Seek`)

Reply:
It makes sense for `FrameDecoder` to have only `Read`, because it's implementing a stream out (while `Seek` would be really unnatural).

However, in principle `.npy` files don't need `Seek`, it's only for `.npz`:
https://docs.rs/ndarray-npy/latest/ndarray_npy/trait.ReadNpyExt.html

Reply:
mh? but we are reading npz
https://github.com/NNPDF/eko/blob/c6784d9bd9274af843d4f4ae384d3f40d0c8cbe3/crates/dekoder/src/inventory.rs#L71

Reply:
ehm - I just use an eko which I most likely created when merging the many PR some time ago as I running the LO LHA several times, so I believe my test object is v0.15 (i.e. master); again I copied the code blindly from PineAPPL and it works - I believe
https://github.com/NNPDF/eko/blob/c81bb08b1f219cf010cf1764b8257925b0eab891/crates/dekoder/tests/test_load.rs#L83
note that I have a single test object and with that I'm happy for the moment :see_no_evil:

Reply:
Can you check the content of the EKO you're testing? Just run `tar -t myeko.tar`.

Reply:
I was about to say, that the problem is with `.npz` and I don't know if I can read that - but yes I can
![grafik](https://github.com/user-attachments/assets/11664c57-2fbf-40b5-bfa6-8b668c90c0ad)
and the thing is really called `operator.npy` - and it has the errors

Reply:
I was only interested in the previous file, the `EFM...ks=.npz`. The one inside should be an arbitrary string, so it doesn't have to have the `.npy` extension at all...

Impressive that the GUI Gnome utility is aware of `.npz` files.

Reply:
It seems like it is an internal implementation detail:

> The .npz file format is a zipped archive of files named after the variables they contain. The archive is not compressed and each file in the archive contains one variable in .npy format.

Being a zip, it explains why Gnome can open it.

I wonder whether we should stop using `.npz` files, and just use the `<hash>.npy` and `<hash>-err.npy` to avoid one archive layer...

This would also be a breaking change, so it should be scheduled for later.

In any case, pay attention that your implementation is making use of a (documented) implementation detail. And it would fail if there is no error around, because in that case the EKO would just contain `<hash>.npy` directly.
Sorry for the confusion, I had no idea of how an `.npz` was implemented at the time (I could read the note, most likely it was already there...), and it seemed a good idea to keep them together. Not that good, in the hindsight...

Reply:
> so it should be scheduled for later.

yes, as said already: I'm expecting at the end of this journey a breaking change and we can break even further

> In any case, pay attention that your implementation is making use of a (documented) implementation detail. And it would fail if there is no error around, because in that case the EKO would just contain `<hash>.npy` directly.

yes, but for the moment I'm happy with my single test case - this is just the first step.

Reply:
in the end do you want me to change something?